### ISSUE TITLE
Possible delays in execution of cells due to delays in updates to cell output & metadata

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode#107916]
**Based on the current implementation this is how our Python (Jupyter) notebooks work**
Assume we are execute all 10 cells, each with a print statement
* Start executing first cell
	* Update metadata with status to indicate it has started
	* Next update execution order
	* Next update the output
	* Next update the cell run status as successful
* Next proceed to the next cell & execute that cell
* Keep doing this for each cell

**Observations**
* We execute each cell in sequence, one after the other.
	* We cannot send all cells together, as we need to check if a cell fails, if one fails, then we stop executing the rest
* For the next cell to get executed, we need to wait for a lot of VSCode document updates
	* Wait for cell metadata & cell output to get updated.
* Thus the over all execution of a notebook is impacted by how fast/slow the notebook updates work in VS Code.

**Why do we need to wait for cell metadata & output to get updated in each execution**
* We don't nee to wait for cell metadata to get updated.
	* Right now we're awaiting, as we are already awaiting on updates the cell output.
* We need to wait for cell output to get updated as we need the previous cell state when appending strings.
	* E.g. when streaming text, we need to update the existing output.
	* Assume a code block is printing numbers 1-100 & sleeps between each output. 
	* At this point we get the streamed text one at a time, however we update the same output object.
	* To append the new output to the exsting output, we need the existing (previous output). This will only work if we await on the updates. Else when we print number `2`, its possible the cell output has not yet been udpated to contain `1`.
	

**Solutions**
* We (extension authors) can avoid awaiting on these updates
	* E.g. not wait for VS Code to complete the updates to cell metadata & output
	* However we still need to wait for the cell output to get updated.
* We (extension authors) can avoid awaiting on cell metadata (execution order, execution times, etc)
	* & optionally VS Code provides us with an API to append output.
	* This way we don't need to wait for any updates, as we can alway execute some VS Code API in a fire & forget pattern.

**WebView based notebook in Python extension**
* We did come across this similar issue with our Webview based notebooks. 
I.e. the bottleneck of waiting for udpates from UI to complete before proceeding to execute the next cell.
* We solved this by not waiting for the UI to get updated. Turned the messages into a fire & forget.

**Note**
Here are the times to execute an entire notebook with 72 cells, each with a print statement
* Notebook executed in Jupyter takes 0.3s
* Notebook in our existing Python (WebView based) notebooks takes 1.2s
* Notebook using VS Code Notebooks UI takes approx 2.5s on my Mac.
	* If i remove all of the awaits on the updates to notebook metadat & output the times come down to almost 0.6s (big improvement).

Though the improvements might not seem much, the implications of this current design can have bad consequences:
* E.g. user could have a cell that streams output. Each output now waits on updates to the notebook.
* This slows down the time taken for our extension to proceed to the next cell.
If the next cell too streams a lot of output, then the next cells slows down.
In summary, execution of cells can get slowed down when we have streamed output & when extension authors wait on updates to the notebook when using the exiting update methods to update metadata & output.

**Summary**
* One easy solution is for VS Code to provide an API to append output.
& extensions can optionally choose to not await on updates to the metadata.

@rebornix @jrieken /cc

**Comment by jrieken** [comment]
@DonJayamanne Please share you samples and how you measure these things. I do not believe that it takes 2.5 seconds to apply an edit and come back.

**Comment by Sambhrama754** [comment]
Pls don't send me emails.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by jrieken** [comment]
related to https://github.com/microsoft/vscode/issues/107424

**Comment by DonJayamanne** [comment]
> I do not believe that it takes 2.5 seconds to apply an edit and come back.

Its not 2.5 to apply a single edit.
Its around 4 edits per cell.
When we run a cell we have the following edits:
* Update cell status to running
* Update execution order when we get that back
* Update output when we get that back
* Finally update cell status (error/success) based on execution result/output

Multiply this 70times for 70 cells, then the time is exacerbated.
I can shave around 1s by removing all of the edits. This way, each cell executes faster from the extension as we're not waiting for UI updates to proceed to the next cell.

As mentioned earlier we can optimize our extension code to not await on some updates.
But waiting for updating output still exists.

Here's the notebook.
[perf.ipynb.txt](https://github.com/microsoft/vscode/files/5319031/perf.ipynb.txt)

**Comment by jrieken** [comment]
The need to extension to concat previous output bits with new output bits is bogus and wasteful, esp. since we will immediately remove what's already there. So, the idea of append or update output is something we should go after. I see two possibilities:

1. Have something similar to `vscode.OutputChannel` 
2. Allow a `NotebookCellOutput`-object to refer to an previous output-object to which is wants to be appended

The former approach is something well-known in API terms but we must be able to "spy" on the object so that we know when 'appendOutput' or such is called. The API concept for this to spec types as interface only and by offering a way to create an instance of said interface, e.g `createNotebookOutput`. This works well for long-lived objects but is likely cumbersome when the lifetime is short - as it is with notebook output. It is cumbersome because the extension owns the object and therefore must dispose it but the UI and API know/control when the object is of no more interest.

Therefore the second option might be better and simpler: An extension can keep hold of the last output and when creating a new object it would pass it along. It could look like so (building on the new output proposal https://github.com/microsoft/vscode/issues/107424)

```ts
export class NotebookCellOutput {

  readonly outputs: NotebookCellOutputItem[];
  readonly metadata?: Record<string, string | number | boolean>;

  constructor(
    outputs: NotebookCellOutputItem[], 
    metadata?: Record<string, string | number | boolean>, 
    appendTo?: NotebookCellOutput // previous output to which this is appended
  );
}
```

This makes it data-only objects which is generally simpler and preferred. What remains is how to signal to an extension that a complete new output is now active? The events we have for that might be enough tho

**Comment by DonJayamanne** [comment]
Thanks for the feedback, I think we can close this issue. 
I'll look for changes to the output and try the existing object model.

**Comment by jrieken** [comment]
merging this then with https://github.com/microsoft/vscode/issues/107424
